pkg/controllers/util/naming.go

```diff
@@ -51,3 +57,43 @@ func qualifiedNameForCluster(clusterName string, qualifiedName common.QualifiedN
 // QualifiedNameForCluster returns the qualified name to use for the
 // given cluster.
 var QualifiedNameForCluster = qualifiedNameForCluster
+
+// GenerateFederatedObjectName will generate name for federated object by ftc name and source object name
+func GenerateFederatedObjectName(objectName, ftcName string) string {
+	transformedName, transformed := transformObjectName(objectName)
+	federatedName := fmt.Sprintf("%s-%s", transformedName, ftcName)
+	if transformed {
+		federatedName = fmt.Sprintf("%s-%d", federatedName, fnvHashFunc(objectName))
+	}
+
+	if len(federatedName) > common.MaxFederatedObjectNameLength {
+		federatedName = federatedName[:common.MaxFederatedObjectNameLength]
+	}
+
+	return strings.ToLower(federatedName)
+}
+
+func transformObjectName(objectName string) (string, bool) {
+	transformed := false
+	if strings.Contains(objectName, ":") {
```

Comment:
nit: we could collapse the 3 checks into a loop over ":$%"

Reply:
Are we sure these are the only possible special characters? Could we insert a link to the apiserver code that defines permitted characters?